modules/renter/contractor/contractor.go

```diff
@@ -0,0 +1,154 @@
+package contractor
+
+import (
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+var (
+	errNilCS           = errors.New("cannot create contractor with nil consensus set")
+	errNilWallet       = errors.New("cannot create contractor with nil wallet")
+	errNilTpool        = errors.New("cannot create contractor with nil transaction pool")
+	errUnknownContract = errors.New("no record of that contract")
+)
+
+// a Contract includes the original contract made with a host, along with
+// the most recent revision.
+type Contract struct {
+	IP              modules.NetAddress
+	ID              types.FileContractID
+	FileContract    types.FileContract
+	MerkleRoots     []crypto.Hash
+	LastRevision    types.FileContractRevision
+	LastRevisionTxn types.Transaction
+	SecretKey       crypto.SecretKey
+}
+
+// A Contractor negotiates, revises, renews, and provides access to file
+// contracts.
+type Contractor struct {
+	// dependencies
+	dialer  dialer
+	hdb     hostDB
+	log     logger
+	persist persister
+	tpool   transactionPool
+	wallet  wallet
+
+	allowance     modules.Allowance
+	blockHeight   types.BlockHeight
+	cachedAddress types.UnlockHash // to prevent excessive address creation
+	contracts     map[types.FileContractID]Contract
+	renewHeight   types.BlockHeight // height at which to renew contracts
+
+	mu sync.RWMutex
+}
+
+// Allowance returns the current allowance.
+func (c *Contractor) Allowance() modules.Allowance {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.allowance
+}
+
+// SetAllowance sets the amount of money the Contractor is allowed to spend on
+// contracts over a given time period, divided among the number of hosts
+// specified. Note that Contractor can start forming contracts as soon as
+// SetAllowance is called; that is, it may block.
+func (c *Contractor) SetAllowance(a modules.Allowance) error {
+	// sanity checks
+	if a.Hosts == 0 {
+		return errors.New("hosts must be non-zero")
+	} else if a.Period == 0 {
+		return errors.New("period must be non-zero")
+	}
+
+	// Set the allowance.
+	c.mu.Lock()
+	old := c.allowance
+	c.allowance = a
+	c.mu.Unlock()
+
+	// If this is the first time the allowance has been set, form contracts
+	// immediately.
+	if old.Hosts == 0 {
+		return c.formContracts(a)
+	}
+
+	// Otherwise, if the new allowance is "significantly different" (to be
+	// defined more precisely later), form intermediary contracts.
+	if a.Funds.Cmp(old.Funds) > 0 {
+		// not yet implemented
+		// c.formContracts(diff(a, old))
+	}
```

Comment:
should drop a TODO here so it's easy to find later once you're in 'go back and finish things' mode